packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/TypedSnippets/ModelSerializationExtensionsSnippet.cs

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.Generator.CSharp.ClientModel.Providers;
+using Microsoft.Generator.CSharp.Expressions;
+using Microsoft.Generator.CSharp.Snippets;
+using Microsoft.Generator.CSharp.Statements;
+
+namespace Microsoft.Generator.CSharp.ClientModel.Snippets
+{
+    internal sealed record ModelSerializationExtensionsSnippet(ValueExpression Untyped) : TypedSnippet<ModelSerializationExtensionsProvider>(Untyped)
+    {
+        private static ModelSerializationExtensionsProvider? _provider;
+        private static ModelSerializationExtensionsProvider Provider => _provider ??= new();
+
+        public static readonly ModelReaderWriterOptionsSnippet Wire = Provider.WireOptions;
+
+        public static MethodBodyStatement WriteObjectValue(Utf8JsonWriterSnippet snippet, TypedSnippet value, ValueExpression? options = null)
+            => Provider.WriteObjectValue(snippet, value, options: options);
```

Comment:
If we are going this approach shouldn't we just move the snippet to this class and no longer have it on the provider itself?

Reply:
or maybe we make this static on the provider.
Removing it from the provider will require us to make some of the values (such as method names) public or hard-coded in multiple places
Because of this reason, we made those providers singleton

Reply:
Dapeng is correct. I've made the constants internal within the providers. Since these expressions rely on the provider's `Type` field, I'm not sure if we can make them static methods.

Reply:
Yeah we cannot make it static because we need the generated `CSharpType` which is the reason why these types are singleton in the first place, we will have to construct them everywhere.
in autorest.csharp, we do not have a place that you could get the context variables via static methods or properties (such as typeFactory)

Reply:
@m-nash and I talked offline. We have just duplicated the method names as private constants within the snippets instead of exposing them within the providers.